RIAK-2949: Refactor vnode request records into requests module with factories and getters

paulhenrich opened the pull request:
Just opening this up so thumbs can tell us when we mess something up.

Jira: https://bashoeng.atlassian.net/browse/RIAK-2949

hazen commented:
@paulhenrich Feel free to look at the `kv_ts_merge` branch to see how this might play out in a KV/TS world

Commit:
Refactor request records into request module

Remove definition and usage of the request macros, for example:
* ?KV_PUT_REQ
* ?KV_GET_REQ
* ?KV_LISTKEYS_REQ

Replace with factories and getters in riak_kv_requests. Within
riak_kv_vnode, we replace pattern matching on the record with
dispatch to functions that match on a request_type getter.

Refactor use of case AsyncFolding to use existing function

Less fun: remove unnecessary function application

Inline module name so it is visible to xref

Extract `fold_type_for_query` function

Also, delete `prepare_index_query` function, as it was nothing more than
an identity function used in exactly one place, and had a bad type spec
to boot. While we were at it, deleted the lines in the
`dialyzer.ignore-warnings` file that referenced the now-deleted
`prepare_index_query` function.

lehoff commented on src/riak_kv_vnode.erl:
```diff
@@ -934,36 +958,15 @@ handle_coverage(?KV_LISTBUCKETS_REQ{item_filter=ItemFilter},
+    Bucket = riak_kv_requests:get_bucket(Req),
+    ItemFilter = riak_kv_requests:get_item_filter(Req),
+    Query = riak_kv_requests:get_query(Req),
+    ResultFun = case riak_kv_requests:get_ack_backpressure(Req) of
```
This is not consistent with the way it is done for `kv_listkeys_request` where the result fun is applied to the `Bucket` and `Sender` on the spot before calling the next function.

Since we have both `Bucket` and `Sender` in our hands right now we should apply them immediately.
And once the usage has been unified we should introduce a `result_fun_for_request(Req, Bucket, Sender)` function that returns the right result fun.

lehoff commented on src/riak_kv_vnode.erl:
```diff
@@ -878,41 +839,104 @@ handle_command({get_index_entries, Opts},
+    ItemFilter = riak_kv_requests:get_item_filter(Req),
     %% Construct the filter function
     Filter = riak_kv_coverage_filter:build_filter(ItemFilter),
     BufferMod = riak_kv_fold_buffer,
```
```
BufferMod = riak_kv_fold_buffer,

Buffer = BufferMod:new(BufferSize, result_fun(Sender)),
FoldFun = fold_fun(buckets, BufferMod, Filter, undefined),
FinishFun = finish_fun(BufferMod, Sender),
```

while the use in the FoldFun and FinishFun is fine., it is problematic in the BufferMod:new/3 since it stops xref from spotting the call. 
Please replace that one with the actual atom.

lehoff replied:
Further down we have

```
case AsyncFolding andalso AsyncBackend of
         true ->
             Opts = [async_fold];
         false ->
             Opts = []
     end,
```

At the very least we shouldn't export `Opts` from the case statement, but even better would be to move the case statement into a function, say, `options_for_folding_and_backend/2`

lehoff commented on src/riak_kv_vnode.erl:
```diff
@@ -934,36 +958,15 @@ handle_coverage(?KV_LISTBUCKETS_REQ{item_filter=ItemFilter},
+    handle_coverage_index(Bucket, ItemFilter, Query, FilterVNodes, Sender, State, ResultFun).
 
 -spec prepare_index_query(?KV_INDEX_Q{}) -> ?KV_INDEX_Q{}.
 prepare_index_query(#riak_kv_index_v3{term_regex=RE} = Q) when
```
On line 998 remove the useless `%% Update stats...` comment.

lehoff replied:
Further down we have:

```
 %% @HACK
             %% Really this should be decided in the backend
             %% if there was a index_query fun.
             FoldType = case riak_index:return_body(Query) of
                            true -> fold_objects;
                            false -> fold_keys
                        end,
```

This should be put into a function `fold_type_for_query/1`

lehoff reviewed (changes requested):
Just minor clean-ups laying around for grabs, when done it can be merged.

hazen commented on include/riak_kv_vnode.hrl:
```diff
@@ -1,83 +1,20 @@
-    % start_time :: non_neg_integer(), Jon to add?
-}).
 
 -record(riak_kv_w1c_put_reply_v1, {
```
Is this record not being refactored because of the same reason as `riak_kv_listkeys_req_v2`?

jvoegele replied:
We left this one as-is because it is a _reply_ record and it would not make sense to mix it into the `riak_kv_requests` module, which (as implied by the name) is intended only for _request_ records. We could, of course, move `riak_kv_w1c_put_reply_v1` into a new module (say `riak_kv_replies`) but since there was only one reply record in existence we decided this was not worth introducing a new module.

hazen commented on src/riak_kv_vnode.erl:
```diff
@@ -1075,32 +1081,27 @@ handle_handoff_command(Req=?KV_PUT_REQ{}, Sender, State) ->
             StartTS = os:timestamp(),
             riak_core_vnode:reply(Sender, {w, Idx, ReqId}),
-            {Reply, UpdState} = do_put(Sender, BKey,  Object, ReqId, StartTime, Options, State),
+            {Reply, UpdState} = do_put(Sender, Req, State),
```
Love how this cleans up the interface

hazen commented on src/riak_kv_vnode.erl:
```diff
@@ -2787,223 +2801,5 @@ rollover_test_() ->
              ]}
     }.
 
-dummy_backend(BackendMod) ->
```
We agreed this was not needed any more, right?

jvoegele replied:
@javajolt Our understanding is that this test code was written for an obsolete testing tool that is no longer in use, and that the test never gets executed.

@bsparrow435 Can you confirm?

hazen commented on test/fsm_eqc_vnode.erl:
```diff
@@ -99,9 +98,14 @@ get_postcommits() ->
-active(?VNODE_REQ{index=Idx,
-                  sender=Sender,
-                  request=?KV_GET_REQ{req_id=ReqId}}, State) ->
+active(?VNODE_REQ{index = Idx,
```
Why are we keeping the macro here?

jvoegele replied:
The `VNODE_REQ` record is defined in `riak_core_vnode.hrl` and since our scope is just the request records defined in `riak_kv_vnode.hrl` we left this one as-is. Per @lehoff this is for another round.

lehoff replied:
I think we need to revisit how we do requests across CORE and KV - the extraction of CORE did not get a super clean separation of concepts.

Now we have the KV side of requests in order and then we can go back and look a the broader picture when we have live with this solution for a while.

lehoff reviewed (approved):
All previous comment addressed.

gordonguthrie commented:
Before we get a merge on this on the KV side I would like an integration branch created and a trial merge with TS done to see if it makes sense to yolomerge this clean-up into TS now or to defer...

hazen commented:
@gordonguthrie You share my thoughts. I've been taking to Jason and Paul about this change already. Would like to simply extend the changes too TS when they are done.

paulhenrich commented:
Thanks for the feedback @gordonguthrie & @javajolt and @lehoff. I agree we should test out the merge. I am currently investigating test failures after a giddyup run and coordinating with Gordon on email.

paulhenrich commented:
I opened PR #1517 to see how the merge might go with TS. It is against a copy of `kv_ts_merge`.

hazen commented on src/riak_kv_vnode.erl:
```diff
@@ -534,17 +520,23 @@ init([Index]) ->
+    riak_core_vnode:reply(Sender, {r, {error, overload}, Idx, ReqId});
+handle_overload_request(kv_w1c_put_request, Req, Sender, _Idx) ->
+    Type = riak_kv_requests:get_replica_type(Req),
     riak_core_vnode:reply(Sender, ?KV_W1C_PUT_REPLY{reply={error, overload}, type=Type});
```
Question previously mentioned: Why is the reply still a macro?

paulhenrich replied:
We are strictly adhering to this being a refactoring of requests in the kv_vnode. It should probably be refactored to follow the same pattern at some point.
